java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,602 @@
+        }
+        break;
+      }
+      case CONTINUE_STATEMENT: {
+        ContinueStatementTree continueStatementTree = (ContinueStatementTree) tree;
+        if(continueStatementTree.label() == null) {
+          if (continueTargets.isEmpty()) {
+            throw new IllegalStateException("'break' statement not in loop or switch statement");
+          }
+          currentBlock = createUnconditionalJump(tree, continueTargets.getLast());
+        } else {
+          currentBlock = createUnconditionalJump(tree, null);
+          gotos.add(currentBlock);
+        }
+        break;
+      }
+      case WHILE_STATEMENT: {
+        WhileStatementTree s = (WhileStatementTree) tree;
+        Block falseBranch = currentBlock;
+        Block loopback = createBlock();
+        // process body
+        currentBlock = createBlock(loopback);
+        continueTargets.addLast(loopback);
+        breakTargets.addLast(falseBranch);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        Block bodyBlock = currentBlock;
+        // process condition
+        currentBlock = createBranch(s, bodyBlock, falseBranch);
+        buildCondition(s.condition(), bodyBlock, falseBranch);
+        loopback.successors.add(currentBlock);
+        currentBlock = createBlock(currentBlock);
+        break;
+      }
+      case DO_STATEMENT: {
+        DoWhileStatementTree s = (DoWhileStatementTree) tree;
+        Block falseBranch = currentBlock;
+        Block loopback = createBlock();
+        // process condition
+        currentBlock = createBranch(s, loopback, falseBranch);
+        buildCondition(s.condition(), loopback, falseBranch);
+        // process body
+        currentBlock = createBlock(currentBlock);
+        continueTargets.addLast(loopback);
+        breakTargets.addLast(falseBranch);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        loopback.successors.add(currentBlock);
+        currentBlock = createBlock(currentBlock);
+        break;
+      }
+      case FOR_EACH_STATEMENT: {
+        ForEachStatement s = (ForEachStatement) tree;
+        //TODO(npe) One solution is to create a forstatement node depending on type of expression (iterable or array) and build CFG from it.
+        Block afterLoop = currentBlock;
+        Block loopback = createBlock();
+        continueTargets.addLast(loopback);
+        breakTargets.addLast(afterLoop);
+        build(s.statement());
+        breakTargets.removeLast();
+        continueTargets.removeLast();
+        build(s.variable());
+        build(s.expression());
+        break;
+      }
+      case FOR_STATEMENT: {
+        ForStatementTree s = (ForStatementTree) tree;
+        Block falseBranch = currentBlock;
+        // process step
+        currentBlock = createBlock();
+        Block stepBlock = currentBlock;
+        for (StatementTree updateTree : Lists.reverse(s.update())) {
```

Comment:
@Godin I don't see the incorrectness in the case you point out  : update blocks are in the correct order AFAIK:

B8:
  0: INT_LITERAL 73aeff7d 0
  1: VARIABLE 6607a2e0 i
  Successors: B7
B7:
  0: IDENTIFIER 7f565474 i
  1: INT_LITERAL 7cc2ccfb 10
  2: LESS_THAN 591fb2ba
  T: FOR_STATEMENT 300eff0c
  Successors: B6 B1
B6:
  0: IDENTIFIER 288a857a i
  1: IDENTIFIER 1e412161 System
  2: MEMBER_SELECT 45f676cb
  3: MEMBER_SELECT 37f12d4f
  4: METHOD_INVOCATION 3ad5ba3a
  Successors: B5
B5:
  0: BOOLEAN_LITERAL 25b55403
  T: CONDITIONAL_EXPRESSION 239ea44c
  Successors: B4 B3
B4:
  0: IDENTIFIER 4348bd9 i
  1: INT_LITERAL 15c955b7 1
  2: PLUS 551a6c8d
  Successors: B2
B3:
  0: IDENTIFIER 39ccc8a1 i
  1: INT_LITERAL 3387dfac 1
  2: PLUS 1d6dafa2
  Successors: B2
B2:
  0: IDENTIFIER c25c3c6 i
  1: ASSIGNMENT 3580265
  Successors: B7
B1:
  Successors: B0
B0 (Exit) :